Framework/Algorithms/test/EditInstrumentGeometryTest.h

```diff
@@ -50,96 +54,329 @@ class EditInstrumentGeometryTest : public CxxTest::TestSuite {
+
+  //----------------------------------------------------------------------------------------------
+  /** Test `EditInstrumentGeometry`:
+   *    verify the layout of the resulting instrument.
+   */
+  void test_InstrumentLayout() {
+    // 1. Init
+    EditInstrumentGeometry editdetector;
+    editdetector.initialize();
+
+    // 2. Load Data
+    DataObjects::Workspace2D_sptr inputWS = WorkspaceCreationHelper::create2DWorkspaceWithFullInstrument(6, 100, false);
+    ADS::Instance().add("inputWS6", inputWS);
+
+    // 3. Set Properties
+    TS_ASSERT_THROWS_NOTHING(editdetector.setPropertyValue("Workspace", "inputWS6"));
+    TS_ASSERT_THROWS_NOTHING(editdetector.setPropertyValue("L2", "1.1, 2.2, 3.3, 4.4, 5.5, 6.6"));
+    TS_ASSERT_THROWS_NOTHING(editdetector.setPropertyValue("Polar", "90.1, 90.2, 90.3, 90.4, 90.5, 90.6"));
+    TS_ASSERT_THROWS_NOTHING(editdetector.setPropertyValue("Azimuthal", "1.0, 2.0, 3.0, 4.0, 5.0, 6.0"));
+
+    // 4. Run
+    TS_ASSERT_THROWS_NOTHING(editdetector.execute());
+    TS_ASSERT(editdetector.isExecuted());
+
+    // 5. Check result
+    Mantid::API::MatrixWorkspace_sptr outputWS;
+    TS_ASSERT_THROWS_NOTHING(
+        outputWS = std::dynamic_pointer_cast<Mantid::API::MatrixWorkspace>(ADS::Instance().retrieve("inputWS6")));
+
+    const auto instrument = outputWS->getInstrument();
+    const auto &detectorInfo = outputWS->detectorInfo();
+    const auto &componentInfo = outputWS->componentInfo();
+
+    // The instrument includes no monitors
+    TS_ASSERT_EQUALS(instrument->getNumberDetectors(true), instrument->getNumberDetectors(false));
+
+    // The instrument has one detector per spectrum
+    TS_ASSERT_EQUALS(detectorInfo.size(), outputWS->getNumberHistograms());
+
+    // The instrument includes a source.
+    TS_ASSERT(componentInfo.hasSource());
+
+    // The workspace includes a sample.
+    TS_ASSERT(componentInfo.hasSample());
+
+    // The component info consists of only these components:
+    //   <detectors> + <detector bank> + <source> + <sample> + <root>
+    TS_ASSERT_EQUALS(componentInfo.size(), detectorInfo.size() + 4);
+
+    std::optional<size_t> bank = std::nullopt;
+    for (size_t index = 0; index < componentInfo.size(); ++index) {
+      // Detectors span a contiguous low-index section of the component info.
+      if (!componentInfo.isDetector(index))
+        break;
+
+      // No detector has the root as its immediate parent.
+      // (Changed from previous: this allows `SaveNexusESS` to actually save the instrument.)
+      TS_ASSERT(componentInfo.parent(index) != componentInfo.root());
+
+      // There is only one detector bank.
+      if (!bank) {
+        bank = componentInfo.parent(index);
+        // The detector bank's immediate parent is the root.
+        TS_ASSERT_EQUALS(componentInfo.parent(*bank), componentInfo.root());
+      } else
+        // Each detector belongs to the same detector bank.
+        TS_ASSERT_EQUALS(componentInfo.parent(index), *bank);
+    }
```

Comment:
is a 'bank' still considered a detector?
Does the bank actually come first in the component order? (l351)
According to l367 shouldnt l362 fail before you arrive at l368 which asserts the opposite?

Reply:
What I'm checking here is that the structure consists of only one bank (who's parent is the root), and that there aren't any _nested_ banks, or other complicated geometry.  This test is mostly to catch the case where someone modifies `EditInstrumentGeometry` _again_, and isn't aware of how it's already being used.
`ComponentInfo` tree is saved (indexing in reverse),  from _root_, _source_ / _sample_, then any banks, tubes, and / or etc, and finally all of the pixels and monitors.
Finally, `isDetector` is kind-of misnamed.  Banks are labeled with `NX_class: NXdetector`.  So the sequence (reverse indexing) guarantees that we see the bank first.

Reply:
In answer to your crossed out questions:  (1) "bank" would have a higher index than all of the pixels, (2) "source" "sample" and "root" would have higher indices still.